File: smart_travel_BE/src/main/java/com/example/smart_travel_BE/entity/RoomType.java
package com.example.smart_travel_BE.entity;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

@Entity
@Table(
        name = "room_types",
        indexes = {
                @Index(name = "idx_hotel_id", columnList = "hotel_id"),
                @Index(name = "idx_price", columnList = "price")
        }
)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class RoomType {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 150)
    private String name;

    private Integer capacity;

    @Column(name = "price", nullable = false, precision = 10, scale = 2)
    private BigDecimal price;

    @Column(name = "total_rooms")
    private Integer totalRooms;

    @Column(name = "available_rooms")
    private Integer availableRooms;

    @Column(columnDefinition = "text")
    private String amenities; // JSON hoặc text, ví dụ: ["WiFi","Breakfast","Pool"]

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "hotel_id")
    private Hotel hotel;

    @OneToMany(mappedBy = "roomType", cascade = CascadeType.ALL)
    private List<Booking> bookings;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (availableRooms == null) {
            availableRooms = totalRooms != null ? totalRooms : 0;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: smart_travel_FE/lib/domain/entities/room_type.dart
class RoomType {
  final int id;
  final String name;
  final int? capacity;        // Sức chứa tối đa
  final double? price;        // Giá / đêm
  final int? availableRooms;  // Số phòng trống (nếu có)
  final List<String>? amenities; // ["WiFi", "Breakfast"]

  RoomType({
    required this.id,
    required this.name,
    this.capacity,
    this.price,
    this.availableRooms,
    this.amenities,
  });

  factory RoomType.fromJson(Map<String, dynamic> json) {
    return RoomType(
      id: json['id'],
      name: json['name'],
      capacity: json['capacity'],
      price: (json['price'] != null)
          ? (json['price'] is int ? (json['price'] as int).toDouble() : json['price'])
          : null,
      availableRooms: json['availableRooms'],
      amenities: json['amenities'] != null
          ? List<String>.from(json['amenities'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'capacity': capacity,
      'price': price,
      'availableRooms': availableRooms,
      'amenities': amenities,
    };
  }
}
